### PR TITLE
Speed up builds of freshly checked out branches

### PR DESCRIPTION
**Pull request by nineinchnick**
Add a `quick` profile, which allows to quickly build the project without any checks or tests, assuming it has just been checked out from a clean branch (master) which has been tested by the CI workflow and is known to be correct. The new profile is supposed to make it easier, instead of having to remember all the correct properties (compare the commands in the next section).

I'm not disabling any modules (like docs) on purpose, they still have to be disabled explicitly. Note that building the tarball with Provisio takes up ~25% of wall-time, and the GC and JIT take ~50%. 

### Timings

I ran all the commands on my macbook with the profiler enabled:
```
export MAVEN_OPTS="-agentpath:$HOME/Downloads/async-profiler-2.5.1-macos/build/libasyncProfiler.dylib=start,event=cpu,file=mvn-profile.html"
```

#### Base

Running on master (only skip tests and checks):

```
./mvnw clean install --strict-checksums -V -DskipTests -Dair.check.skip-all -pl '!:trino-server-rpm,!docs' 
```

yields `Total time:  04:41 min`.

When also skipping test compilation, building source JARs, and others:

```
./mvnw clean install --strict-checksums -V -DskipTests -Dmaven.test.skip=true -Dmaven.source.skip=true -Dmaven.site.skip=true -Dmaven.javadoc.skip=true -Dair.check.skip-all -pl '!:trino-server-rpm,!docs'
```

yields `Total time:  03:48 min`.

Running the same command using Maven Daemon, the 3rd run after starting the daemon yields `Total time:  01:28 min (Wall Clock)`.

#### With the profile

Running:
```
./mvnw clean install --strict-checksums -V -Dquick -pl '!:trino-server-rpm,!docs'
```
yields `Total time:  03:08 min`, 40 seconds quicker.

Running this 3rd time after starting the daemon:
```
mvnd clean install --strict-checksums -V -Dquick -pl '!:trino-server-rpm,!docs'
```

yields `Total time:  01:19 min (Wall Clock)`, 10 seconds quicker.

All the work here is based solely on https://peter.palaga.org/presentations/211021-baselone-maven-my-life-is-short/

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1848,5 +1858,142 @@
...
+        <profile>
+            <id>quick</id>
+            <activation>
+                <property>
+                    <name>quick</name>
+                </property>
+            </activation>
+            <properties>
+                <skipTests>true</skipTests>
+                <maven.test.skip>true</maven.test.skip>
```
Comment (hashhar) [inline pom.xml:1870]: test compilation skipping is not useful in practice IMO. We change as much test code as product code in each PR.

In my local bash aliases I skip the source and javadoc steps though.

File: pom.xml [hunk 2]
```diff
@@ -1848,5 +1858,142 @@
...
+                <maven.site.skip>true</maven.site.skip>
+                <maven.source.skip>true</maven.source.skip>
+                <maven.javadoc.skip>true</maven.javadoc.skip>
+                <air.check.skip-all>true</air.check.skip-all>
+            </properties>
+            <build>
+                <plugins>
+                    <!-- Disable plugins entirely, instead of setting the skip option -->
```
Comment (hashhar) [inline pom.xml:1878]: What's time-delta between skip-all vs disabling plugins entirely? We need to keep this list in sync with any new checks that get added in airbase and it's quite verbose.

**Review by hashhar — COMMENTED** [review]
Who is the expected user of the quick profile?

Can you extract "Run git commit id Maven plugin only once" to separate PR. It's non-controversial and always a gain so it can be merged right now.

File: pom.xml [hunk 3]
```diff
@@ -1743,6 +1743,16 @@
...
                     <version>1.0.18</version>
                 </plugin>
+
+                <plugin>
+                    <groupId>pl.project13.maven</groupId>
+                    <artifactId>git-commit-id-plugin</artifactId>
+                    <configuration>
+                        <runOnlyOnce>true</runOnlyOnce>
+                        <injectAllReactorProjects>true</injectAllReactorProjects>
+                        <offline>true</offline>
```
Comment (findepi) [inline pom.xml:1753]: What does offline mean?
Can you extract this change to separate PR to have it merged faster?
seems orthogonal to quick-related considerations

File: pom.xml [hunk 4]
```diff
@@ -1848,5 +1848,142 @@
...
+                        <executions>
+                            <execution>
+                                <id>default</id>
+                                <phase>none</phase>
+                            </execution>
+                        </executions>
+                    </plugin>
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-pmd-plugin</artifactId>
```
Comment (findepi) [inline pom.xml:1932]: not used?

File: pom.xml [hunk 5]
```diff
@@ -1848,5 +1848,142 @@
...
+                        <executions>
+                            <execution>
+                                <id>default</id>
+                                <phase>none</phase>
+                            </execution>
+                        </executions>
+                    </plugin>
+                    <plugin>
+                        <groupId>org.jacoco</groupId>
+                        <artifactId>jacoco-maven-plugin</artifactId>
```
Comment (findepi) [inline pom.xml:1952]: not used in the project, but still initialized by airbase only in order to be skipped.

maybe we could have some airbase change not to pull plugins that are only there to be skipped?
we could start with those easier once (jacoco, pmd), to avoid problems like https://github.com/airlift/airbase/pull/292

File: pom.xml (hunk 3 above)
Comment (hashhar) [reply pom.xml:1753]: @nineinchnick `offline` might not be needed since Airbase explicitly lists the properties it needs and none of them need a remote fetch. See https://github.com/airlift/airbase/pull/245/files

File: pom.xml [hunk 6]
```diff
@@ -1848,5 +1848,142 @@
                 </plugins>
             </build>
         </profile>
+        <profile>
+            <id>quick</id>
+            <activation>
+                <property>
+                    <name>quick</name>
```
Comment (findepi) [inline pom.xml:1865]: CI builds could use this for their MAVEN_FAST_INSTALL opts.

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
cc @losipiuk

File: pom.xml (hunk 1 above)
Comment (kokosing) [reply pom.xml:1870]: I think we should use this profile on the CI to compile before tests execution. That way it will be useful and possibly we could simplify `MAVEN_FAST_INSTALL` env variable.

Also we could possibly extract a shell script that could be used do a quick build with all the magic (maybe with some caching in future).

Comment (hashhar) [reply pom.xml:1870]: caching on CI is already here since https://github.com/trinodb/trino/pull/10115.

Agreed about MAVEN_FAST_INSTALL. 👍

File: pom.xml (hunk 2 above)
Comment (hashhar) [reply pom.xml:1878]: I should read more. It's 40 seconds without mvnd. 10 seconds with mvnd.

File: pom.xml (hunk 1 above)
Comment (nineinchnick) [reply pom.xml:1870]: The goal here is to be able to run the server as quickly as possible. Compiling tests is not required for this. It's useful in some CI jobs that separate running checks and running tests, to detect compilation failures earlier, but I think we should skip them here.

File: pom.xml (hunk 5 above)
Comment (nineinchnick) [reply pom.xml:1952]: Yeah, I realized it would make more sense to add this profile in airbase. I'm not sure that disabling them this way has _all_ the benefits.

File: pom.xml (hunk 1 above)
Comment (hashhar) [reply pom.xml:1870]: Got it. So `maven-checks` does the full checks so that other jobs can skip them and finish faster.

File: pom.xml (hunk 3 above)
Comment (nineinchnick) [reply pom.xml:1753]: https://github.com/trinodb/trino/pull/10521

File: pom.xml (hunk 6 above)
Comment (nineinchnick) [reply pom.xml:1855]: Done!

File: pom.xml (hunk 2 above)
Comment (nineinchnick) [reply pom.xml:1868]: The 40 seconds difference is most likely caused by skipping test compilation, not just disabling plugins. I ran this once now and the difference gets smaller, around 10 seconds, for only disabling plugins, and since I'm running this on my MacBook the error here might be pretty high, so I'd need to run this multiple times and it's 4 minutes each time, so it'll take me a while.

Comment (nineinchnick) [reply pom.xml:1868]: Anyway, I'll look into adding this to airbase, and measure again.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -16,7 +16,7 @@ env:
   # maven.wagon.rto is in millis, defaults to 30m
   MAVEN_OPTS: "-Xmx512M -XX:+ExitOnOutOfMemoryError -Dmaven.wagon.rto=60000"
   MAVEN_INSTALL_OPTS: "-Xmx2G -XX:+ExitOnOutOfMemoryError -Dmaven.wagon.rto=60000"
-  MAVEN_FAST_INSTALL: "-B --strict-checksums -V --quiet -T C1 -DskipTests -Dair.check.skip-all"
+  MAVEN_FAST_INSTALL: "-B --strict-checksums -V --quiet -T C1 -Dquick -Dmaven.test.skip=false"
```
Comment (findepi) [inline .github/workflows/ci.yml:19]: document why `maven.test.skip=false`, it's not intuitive

per https://github.com/trinodb/trino/pull/10519#discussion_r781284002, it seems necessary to retrigger downloads

also i just realized that, since `-Dquick` skips whole plugins altogether, it's will actually skip downloading their dependencies. This means the build will be faster, but subequent `mvn test` may fail due to dep download.

**Review by findepi — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 7 above)
Comment (nineinchnick) [reply .github/workflows/ci.yml]: I reverted this change.

File: pom.xml (hunk 1 above)
Comment (nineinchnick) [reply pom.xml:1860]: I was considering disabling some modules in this profile, but it would require splitting up the list of modules so that some are added in a default profile _unless_ quick is enabled. Would that be ok? Or should we discuss this in a follow-up PR? The benchmark and verifier modules take 7 seconds to build. I excluded docs and rpm from all my measurements anyway.

@kokosing the point of having this profile is that you don't need any other Maven flags, other than module selectors, so maybe a script is not necessary?

File: pom.xml (hunk 6 above)
Comment (findepi) [reply pom.xml:1855]: taken back per https://github.com/trinodb/trino/pull/10519#discussion_r781285821

**Review by findepi — APPROVED** [review]

File: pom.xml (hunk 1 above)
Comment (findepi) [reply pom.xml:1860]: > so maybe a script is not necessary?

i hope we don't end up with a script. 

Script works great for a silo'ed command, one that you never need to modify, which is not what engineers need on the CLI. I sometimes need to build one module, sometimes two, sometimes the whole project. 

> I was considering disabling some modules in this profile, but it would require splitting up the list of modules so that some are added in a default profile unless quick is enabled. Would that be ok? Or should we discuss this in a follow-up PR? The benchmark and verifier modules take 7 seconds to build. I excluded docs and rpm from all my measurements anyway.

Let's follow up.

File: pom.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (nineinchnick) [reply pom.xml:1855]: Let's only use a profile with `-Pquick` since it's less ambiguous, especially for users not familiar with Maven.

File: pom.xml (hunk 1 above)
Comment (losipiuk) [reply pom.xml:1860]: Question: would that be still possible to do `-Pquick -Dmaven.test.skip=false` to have all the benefits of quick compilation - but still compile test code too?

File: pom.xml (hunk 5 above)
Comment (nineinchnick) [reply pom.xml]: https://github.com/airlift/airbase/pull/293

File: pom.xml (hunk 1 above)
Comment (nineinchnick) [reply pom.xml]: I had to split this into two separate profiles:
* quick - disables checks, javadoc, site and sources
* noTests - doesn't compile tests and disable the dependency plugin

Without disabling the dependency plugin, build fails when `~/.m2/repository` is empty since dependency analysis is executed even when tests are not being built and requires artifacts from all scopes to be resolvable.

I'll resolve this thread since the change is outdated. Please comment on the new profiles if you think they're useful or not. We don't need to keep them in the project, devs could define them in `%USER_HOME%/.m2/settings.xml`

Comment (nineinchnick) [reply pom.xml]: Also, I'll update the PR description with new benchmarks.

Comment (findepi) [reply pom.xml]: > I had to split this into two separate profiles:
> 
> quick - disables checks, javadoc, site and sources
> noTests - doesn't compile tests and disable the dependency plugin

Sounds like a complication. Do we really need it?

In the worst case, there will be occasions where `-Pquick` cannot be used, right?

Does this also mean that `-Pquick -PnoTests` is faster than `-Pquick` alone?

Comment (nineinchnick) [reply pom.xml]: Aaand turns out that the shade plugin also triggers full dependency resolution, so there's no way of actually fully disabling test compilation.

**Review by kokosing — APPROVED** [review]

**Comment by nineinchnick** [comment]
I'm closing this. The gain of adding options other than `-DskipTests -Dair.check.skip-all` is tiny, about 4 seconds, and we can't reliably disable test compilation. Other changes are in https://github.com/airlift/airbase/pull/293

**Comment by nineinchnick** [comment]
Benchmark results:
```
Benchmark 1: ./mvnw clean install -Pquick -pl "!:trino-server-rpm,!docs"
  Time (mean ± σ):     382.767 s ± 28.953 s    [User: 509.683 s, System: 24.475 s]
  Range (min … max):   362.434 s … 415.916 s    3 runs
 
Benchmark 2: ./mvnw clean install -DskipTests -Dair.check.skip-all=true -pl "!:trino-server-rpm,!docs"
  Time (mean ± σ):     368.216 s ±  1.681 s    [User: 512.733 s, System: 25.812 s]
  Range (min … max):   366.666 s … 370.003 s    3 runs
```
There's some variance for `-Pquick` but notice how small is difference in the minimum time.

I also ran benchmarks for the `noTests` profile, that only works when `~/.m2/repository` is populated (any previous `mvn install` completed):
```
Benchmark 1: ./mvnw clean install -Pquick,noTests -pl "!:trino-server-rpm,!docs"
  Time (mean ± σ):     276.097 s ±  0.659 s    [User: 366.680 s, System: 21.953 s]
  Range (min … max):   275.593 s … 276.842 s    3 runs
```

**Comment by ksobolew** [comment]
Is there some value in the first commit alone? Seems like a nice cleanup at least.

**Comment by nineinchnick** [comment]
It'll allow to avoid loading unused plugins in some cases, but the actual time difference is 1-2 second. I can open up another PR with only that if https://github.com/airlift/airbase/pull/293 gets merged, there won't be any difference without it.
